sam/app/template.yaml

```diff
@@ -57,6 +64,13 @@ Parameters:
     Description: GitHub repo branch name. It defaults to master if not specified.
     Type: String
     Default: master
+  CodeCommitRepoName:
+    Type: String
+    Description: CodeCommit repository name, only specify if you chose CodeCommit in SourceCodeProvider
+  CodeCommitBranch:
+    Type: String
+    Description: CodeCommit repository branch name, only specify if you chose CodeCommit in SourceCodeProvider.
+    Default: master
```

Comment:
The combination of required/optional parameters don't make sense. I'm assuming you would configure the app to use either GitHub or CodeCommit and not both, right? If that's the case, it doesn't make sense for both GitHub and CodeCommit parameters to be required. Also, GitHubBranch and CodeCommitBranch could be consolidated into a single Branch parameter, although that's not backwards-compatible so we'd need to major version bump the app.

I think we should make the GitHub and CodeCommit parameters optional and then I think we could define [rules](https://docs.aws.amazon.com/servicecatalog/latest/adminguide/reference-template_constraint_rules.html) in the template to ensure either one or the other is filled in. I haven't actually used Rules before, but my understanding is even though they're documented in Service Catalog's docs, they're actually implemented in CloudFormation so could still be used.

Reply:
Ah, interesting. I wasn't aware of the Rules feature. Let me give it a try and I'll update the PR.

Reply:
@jlhood A couple of lessons learned about using CloudFormation Rules: 

1) You cannot reuse the Conditions defined in the _Conditions_ section, so I couldn't use my `UseCodeCommit` and `UseGitHub` conditions, I had to repeat them in the `Rules` section. So there is a little bit of duplication in line #127 and #138. 

2) ChangeSets do not detect assertion errors. You only get the failure until you execute a CloudFormation change set. 

But overall I like it, makes the template more solid. Let me know your thoughts.

Reply:
@ferdingler Thanks for sharing your learnings!